src/plot.js

```diff
@@ -486,73 +530,133 @@ function nolabel(axis) {
+  const fx = x != null ? Channel(data, {value: x, scale: "fx"}) : undefined;
+  const fy = y != null ? Channel(data, {value: y, scale: "fy"}) : undefined;
+  const groups = facetGroups(data, fx, fy);
+  // If the top-level faceting is non-trivial, track the corresponding data
+  // length, in order to compare it for the warning above.
+  const dataLength =
+    groups.size > 1 || (fx && fy && groups.size === 1 && [...groups][0][1].size > 1) ? data.length : undefined;
+  return {fx, fy, groups, data: facet.data, dataLength};
 }
 
-// A Map-like interface that supports paired keys.
-class FacetMap2 extends FacetMap {
-  has([key1, key2]) {
-    const map = super.get(key1);
-    return map ? map.has(key2) : false;
+// Returns the facet groups, and possibly fx and fy channels, associated with a
+// mark, either through top-level faceting or mark-level facet options {fx, fy}.
+function maybeMarkFacet(mark, topFacetState) {
+  if (mark.facet === null) return;
+
+  // This mark defines a mark-level facet.
+  const {fx: x, fy: y} = mark;
+  if (x != null || y != null) {
+    const data = arrayify(mark.data);
+    if (data == null) throw new Error(`missing facet data in ${mark.ariaLabel}`);
+    // TODO Call applyScaleTransforms on the fx and fy channels.
+    const fx = x != null ? Channel(data, {value: x, scale: "fx"}) : undefined;
+    const fy = y != null ? Channel(data, {value: y, scale: "fy"}) : undefined;
+    return {fx, fy, groups: facetGroups(data, fx, fy)};
   }
-  get([key1, key2]) {
-    const map = super.get(key1);
-    return map && map.get(key2);
-  }
-  set([key1, key2], value) {
-    const map = super.get(key1);
-    if (map) map.set(key2, value);
-    else super.set(key1, new InternMap([[key2, value]]));
-    return this;
+
+  // This mark links to a top-level facet, if present.
+  if (topFacetState === undefined) return;
+
+  const {groups, data, dataLength} = topFacetState;
+  if (mark.facet !== "auto" || mark.data === data) return {groups};
+
+  // Warn for the common pitfall of wanting to facet mapped data. See above for
+  // the initialization of dataLength.
+  if (dataLength !== undefined && arrayify(mark.data)?.length === dataLength) {
+    warn(
+      `Warning: the ${mark.ariaLabel} mark appears to use faceted data, but isn’t faceted. The mark data has the same length as the facet data and the mark facet option is "auto", but the mark data and facet data are distinct. If this mark should be faceted, set the mark facet option to true; otherwise, suppress this warning by setting the mark facet option to false.`
+    );
   }
 }
+
+// Facet filter, by mark; for now only the "eq" filter is provided.
+function filterFacets(facets, {fx, fy}) {
+  const X = fx?.value;
+  const Y = fy?.value;
+  const index = range(X || Y);
+  const gx = X && group(index, (i) => X[i]);
+  const gy = Y && group(index, (i) => Y[i]);
+  return X && Y
+    ? facets.map(({x, y}) => arrayify(intersection(gx.get(x) ?? [], gy.get(y) ?? [])))
```

Comment:
I must be tired but I’m not following why we need to use intersection here… I feel like there’s some redundancy between this calculation at the nested groups we construct in facetGroups. I’m not following why we have to do both “grouping” and “filtering” now for facets.

Reply:
Yes. I changed it back to using the nested groups.

Explanation: this change was a remnant from the work on cumulative facets filter. When the filter is not "eq" but, say, "lte", we need to iterate over the whole cross product of gx x gy. But that's for a future iteration.